language/move-lang/src/errors/diagnostic_codes.rs

```diff
@@ -0,0 +1,105 @@
+// Copyright (c) The Diem Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use codespan_reporting_new::diagnostic::Severity;
+
+//**************************************************************************************************
+// Main types
+//**************************************************************************************************
+
+#[derive(PartialEq, Eq, Clone, Debug, Hash)]
+pub(crate) struct DiagnosticInfo {
+    pub(crate) severity: Severity,
+    category: Category,
+    code: u8,
+    message: &'static str,
+}
+
+pub(crate) trait DiagnosticCode: Copy {
+    const SEVERITY: Severity;
+    const CATEGORY: Category;
+
+    fn code_and_message(self) -> (u8, &'static str);
+
+    fn into_info(self) -> DiagnosticInfo {
+        let severity = Self::SEVERITY;
+        let category = Self::CATEGORY;
+        let (code, message) = self.code_and_message();
+        DiagnosticInfo {
+            severity,
+            category,
+            code,
+            message,
+        }
+    }
+}
+
+//**************************************************************************************************
+// Categories and Codes
+//**************************************************************************************************
+
+macro_rules! codes {
+    ($(severity:$sev:ident category:$cat:ident=$cat_num:literal name:$name:ident codes:[
+        $($code:ident = $code_num:literal: $code_msg:literal),*
+
+    ]),*) => {
+        #[derive(PartialEq, Eq, Clone, Copy, Debug, Hash)]
+        #[repr(u8)]
+        pub enum Category {
+            $($cat = $cat_num,)*
+        }
+
+        $(
+            #[derive(PartialEq, Eq, Clone, Copy, Debug, Hash)]
+            #[repr(u8)]
+            pub enum $name {
+                $($code = $code_num,)*
+            }
+
+            impl DiagnosticCode for $name {
+                const SEVERITY: Severity = Severity::$sev;
+                const CATEGORY: Category = Category::$cat;
+
+                fn code_and_message(self) -> (u8, &'static str) {
+                    let code = self as u8;
+                    debug_assert!(code > 0);
+                    let message = match self {
+                        $(Self::$code => $code_msg,)*
+                    };
+                    (code, message)
+                }
+            }
+        )*
+
+    };
+}
+
+codes!(
+    severity: Error category: NameResolution=2 name: NameResolutionError codes: [
```

Comment:
Loving this macro! However it feels weird to not have commas separating these things -- could you possibly add that in?


Also not needed right now, but I'm wondering if there's a way to check the `cat_num`s to make sure they're all <= 99?